gh ext browse

vilmibm opened the pull request:
Fixes #6302

a new command. a TUI for browsing gh extensions. uses `tview`.

I've tested this pretty extensively on mac, WSL, and Windows (windows terminal + powershell), but more QA is very welcome given the sensitivity of a UI like this in different terminal environments.

there is a [video demo](https://youtu.be/TNjp0ohI7UU)

## discussion points

- The code is not quite as organized as I would like but I've been working on this too long and just need to call it.
- the ReadmeGetter thing is mostly vestigial and could probably be inlined but I ran out of steam
- The test coverage is pretty good but not 100%. It felt like diminishing returns, but it's possible to further inject the dependencies around `tview.Application`

followed by #6540

Commit:
gh ext browse

vilmibm commented:
- [x] use IsStdoutTTY not CanPrompt
- [x] filter out anything that doesn't start with 'gh-'

samcoe commented:
I took this for a spin a bit today and overall functionality worked well. My biggest concern is the colors.


<img width="1791" alt="ext-browse" src="https://user-images.githubusercontent.com/7969779/200300518-9282eddd-baa0-4492-b689-b5fe2e7600fc.png">


This is what it looks like in my normal terminal. It was very difficult to read and use. 


<img width="1776" alt="ext-browse2" src="https://user-images.githubusercontent.com/7969779/200300743-459ba655-1e54-46e9-a571-8239746d5445.png">


This is what is looks like in the default terminal app. Much better. 


Clearly the issue is that my color scheme in my normal terminal has modified colors causing these difficulties. I think this is a common case where color schemes override colors and thus we cant control the experience for all our users. This made me think that we could put in some effort into making the display of the command agnostic of the terminal theme. Right now we specify colors using `tcell`, for example `tcell.ColorWhite`, which will use the terminals definition of that color (only for the base 16 terminal colors), but `tcell` also allows specifying colors by hex value, `tcell.Color(0xFFFFFF)` which will then not rely on the terminal color scheme. What are your thoughts on making our own hex code definitions for the colors we use so that they are consistently displayed?

mislav commented on pkg/cmd/extension/browse/browse.go:
```diff
@@ -0,0 +1,503 @@
+
+func ExtBrowse(opts ExtBrowseOpts) error {
+	if opts.Debug {
+		f, err := os.CreateTemp("/tmp", "extBrowse-*.txt")
```
If the first argument is an empty string, then CreateTemp will create a log file in the system's default temporary location, which I think would be more portable across platforms

mislav commented on pkg/cmd/extension/browse/browse.go:
```diff
@@ -0,0 +1,503 @@
+		return err
+	}
+
+	cacheTTL, _ := time.ParseDuration("24h")
```
For static duration there is no need to parse a human-readable duration: `time.Hour * 24`

mislav commented on pkg/cmd/extension/browse/browse.go:
```diff
@@ -0,0 +1,503 @@
+		opts.Logger = log.New(io.Discard, "", 0)
+	}
+
+	opts.IO.StartProgressIndicator()
```
Maybe this could use an addition of `defer StopProgressIndicator()` so that the progress indicator is made to stop spinning in case of an error

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,6 +228,81 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+						return err
+					}
+
+					io.Out = f
+					io.ErrOut = ioutil.Discard
```
Why is `IOStreams.Out` and `ErrOut` overriden? This disables `StartProgressIndicator`, which is only visible if both streams are connected to a terminal

vilmibm replied:
I've moved this redirection and also limited it to just IO.ErrOut.

I'm stuck doing this because the extension manager's invocation of `git` prints to stderr during an extension install which was corrupting the TUI.

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,6 +228,81 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+						return err
+					}
+
+					cacheTTL, _ := time.ParseDuration("24h")
```
`time.Hour * 24`

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -220,6 +228,81 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					cacheTTL, _ := time.ParseDuration("24h")
+					searcher := search.NewSearcher(api.NewCachedHTTPClient(client, cacheTTL), host)
+
+					f, err := os.Open("/dev/null")
```
Does opening `/dev/null` work on Windows?

vilmibm replied:
it does (but as I said above it's only err out that i needed to deal with)

mislav reviewed (dismissed):
This looks great! Love the work that went into this and the attention to detail. The help docs are absolutely great.

Some results of my own QA in order of what I consider important:

1. The `gh ext browse` command on first run seems to be stuck for about ~20 seconds, but in fact it is preloading data. This is because the spinner isn't visible; see line comments for more notes on this.
2. The lines of the README preview seem double-spaced, which looks wrong (see screenshot). My terminal width is 158.
3. README preview has broken-looking line wrapping (see screenshot).
4. README preview can be scrolled up/down with <kbd>Page Up</kbd>/<kbd>Page Down</kbd>, but I'm not sure if laptop users know how to find that on a keyboard that seemingly doesn't have those keys. On a MacBook Pro, the trick is to press <kbd>Fn</kbd> + <kbd>Up</kbd>/<kbd>Down</kbd> arrows.
5. The spacing in the left sidebar that lists repos and their descriptions looks weird when some repos have descriptions but others not (see screenshot).
6. I think the `Filter` bar need not be visible until someone has entered filtering mode or when there is an active filter. Due to its prominent position, it might be distracting otherwise.

<img width="705" alt="Screen Shot 2022-11-07 at 20 33 02" src="https://user-images.githubusercontent.com/887/200406293-e4f2608b-d9cc-4cee-ab11-23da6c665978.png">

<img width="712" alt="Screen Shot 2022-11-07 at 20 34 17" src="https://user-images.githubusercontent.com/887/200406344-76af5c51-4950-462b-92b1-26371bab1236.png">


> Clearly the issue is that my color scheme in my normal terminal has modified colors causing these difficulties. I think this is a common case where color schemes override colors and thus we cant control the experience for all our users.

@samcoe If we always stick to using the primary 8 ANSI colors in our UI (with the exception of Markdown rendering with Glamour), then we shouldn't need to hardcode our hex codes, because the user will always be responsible to adjust their terminal color scheme so that all primary colors are visible. I'm not sure how tview handles primary colors, though. Maybe in tview interfaces it's the right approach to take complete control over the color scheme and not rely on primary ANSI colors?

vilmibm commented:
@mislav

> 1. The gh ext browse command on first run seems to be stuck for about ~20 seconds, but in fact it is preloading data. This is because the spinner isn't visible; see line comments for more notes on this.

I've fixed this. I had to redirect stderr (turns out I didn't need to redirect out also) because the extension manager prints `git` output to stderr during an install. I just do that redirection after the spinner is done, now.

> 2. The lines of the README preview seem double-spaced, which looks wrong (see screenshot). My terminal width is 158.

I'm genuinely unsure how to affect this from looking at glamour's docs

> 3. README preview has broken-looking line wrapping (see screenshot).

Fixed by matching glamour's word wrap with the width of the readme textview.

> 4. README preview can be scrolled up/down with Page Up/Page Down, but I'm not sure if laptop users know how to find that on a keyboard that seemingly doesn't have those keys. On a MacBook Pro, the trick is to press Fn + Up/Down arrows.

I had assumed this was widespread knowledge for mac users (I use fn + up/down arrow all the time) but I might be wrong. I've called it out in the usage, but do you think we should explore another hotkey?

> 5. The spacing in the left sidebar that lists repos and their descriptions looks weird when some repos have descriptions but others not (see screenshot).

I've changed the code to print "no description provided" when description is blank. Not ideal, but will at least make things more visually consistent.

> 6. I think the Filter bar need not be visible until someone has entered filtering mode or when there is an active filter. Due to its prominent position, it might be distracting otherwise.

I waffled on this but feel that having it always visible is a hint that you _can_ filter. I'm worried about people not noticing the legend at the bottom and a tantalizing filter bar gives them a reason to look for it. More technically, I was not able to find an easy way to hide and show things with tview which was surprising so at this point I'm not inclined to hack at that (but would like to figure out good ways to do that in the future).

vilmibm commented:
@samcoe I'm punting on color because I agree with @mislav about the relative safety of sticking with the primary colors as known to ANSI. I have tweaked it to not use purple, however, which I saw as the worst part of your color scheme and generally a color I could see being wildly different across people's terminals. I've switched instead to just using black and white for list titles and selected highlights.

vilmibm commented:
This is ready for further review @mislav @samcoe

samcoe commented on pkg/cmd/extension/browse/browse.go:
```diff
@@ -0,0 +1,527 @@
+	})
+
+	filter.SetDoneFunc(func(key tcell.Key) {
+		switch key {
```
What do you think about allowing the down key to also do `extList.Focus()`? I naturally reached for that key after I was done typing in my search filter.

samcoe reviewed (approved):
This is great! I am excited for this to merge and hear user feedback on it.

LangLangBart commented:
### Key
> > 7\. README preview can be scrolled up/down with Page Up/Page Down, but I'm not sure if laptop users know how to find that on a keyboard that seemingly doesn't have those keys. On a MacBook Pro, the trick is to press Fn + Up/Down arrows.
> 
> I had assumed this was widespread knowledge for mac users (I use fn + up/down arrow all the time) but I might be wrong. I've called it out in the usage, but do you think we should explore another hotkey?

I learned about `Fn + Up/Down` through this discussion. 😅 

In [fzf (Add shift-up and shift-down)](https://github.com/junegunn/fzf/commit/417bca03dfa9a15a37d64678b5a4a175db6eb0a1) the default for `Up/Down` of the preview is `Shift+↑↓`.

---

### Milky
I get a milky display on macOS 10.15 with the default terminal (bottom-left:`iTerm2`, top-right: `Terminal`).

https://user-images.githubusercontent.com/92653266/200499381-c0242b4d-835f-4be4-a8b7-8749e70e232a.mp4

---

### nit-pick - description for the bottom text 
These are just visual preferences for a different layout, which I think is more visually appealing (can be safely ignored by the author of this PR).
#### Repo: [samtay/so](https://github.com/samtay/so) 
<img src="https://user-images.githubusercontent.com/92653266/200503507-aea1932b-3a63-4374-8715-57a53d5939c0.png" width="600">
<img src="https://user-images.githubusercontent.com/92653266/200503516-6c413897-7256-44ae-aae7-cf20ec2d41f9.png" width="600">

#### Repo: [dlvhdr/gh-dash](https://github.com/dlvhdr/gh-dash) 
<img src="https://user-images.githubusercontent.com/92653266/200504293-5449e5d8-721e-4c39-b281-afd1d10ba4f0.png" width="600">
<img src="https://user-images.githubusercontent.com/92653266/200504305-d0bacb8f-69fa-4cb0-9aae-b4749becf1f9.png" width="600">

---

PS: 🥇 A good addition to the cli tool.

vilmibm commented:
@LangLangBart thanks for the feedback!

re: milkiness, it's unfortunate but I think we're stuck with it for now. I actually struggled to make tview set a background color application wide.

re: the help dialog, I do prefer the idea of having toggle-able help. It was an initial goal that I didn't get to.

I've opened up #6584 to capture this stuff (it already has toggle-able help) and some other followup work we'd like to do.
